# Summary:

_The below task will help you revise all the topics that were covered in the previous weeks. Here, you will also understand what is meant by "scope" of a variable._

## RADAR equation:

RADAR (_Radio Detection and Ranging_) sensors are highly reliable sensors that are not affected by bad weather conditions. Due to their reliability, RADARs find their application in automobiles. One of the most important equations associated with RADARs is the **RADAR-equation** which gives the value of **Signal to Noise ratio (SNR)**. The equation is given below,

**SNR = (P_T * G_T * G_R * lambda_r^2 * sigma) / ( (4 * pi)^3 * R^4 * k * T_o * B * F_n * L)**

Here,

1. **P_T** is the peak transmitted power

2. **G_T** is the gain of the transmitting antenna

3. **G_R** is the gain of the receiving antenna

4. **lambda_r** is radar wavelength (lambda = 0.0375 m)

5. **sigma** is the target radar cross-section or RCS (sigma = 3.98 m^2)

6. **R** is RADAR range

7. **k** is Boltzman's constant and **T_o** denotes reference temperature in Kelvin. For the present task consider (k\*T_o = 4\*10^-21 W-s)

9. **B** is effective noise bandwidth (B = 2.5\*10^6 m)

10. **F_n** is RADAR noise figure (F = 6.31 W/W)

11. **L** is loss term

12. **pi** is the mathematical constant. Use the **math-module** for this.


### Task 1:

_This task is meant to cover all topics covered in week-1(variable, values, assignments, operations, operator precedence, expressions and statements) and week-2 (built-in functions, user-defined functions, scope)_

Your job is to create a function that calculates and returns the **SNR value**.  As you can see from the equation above, some terms are constants. These terms can be defined as **local variables**.  And other variables can be defined as **global varaibles** and passed as **keyword arguments**. 

Once the function is defined, calculate the SNR value for the below values;

1. P_T = 10^6 W

2. G_T = 6309.6 W

3. G_R = 6309.6 W

4. R = 60\*10^3 m

5. L = 5.01 W/W

**Hint:** You can directly substitute the values _without_ bothering about the units.

**Hint:** Here **^** is "to the power of". So, **10^4** is read as **10 to the power of 4**.


In [3]:
# Import math module here to use math.pi

import math

In [4]:
# Your code below:
# -------------------
# NOTE: (5-10 min to solve this task)

def signal_to_noise(sigma, p_t, g_t, g_r, radar_range, loss):
    lambda_r = 0.0375
    k_to = 4*10**-21
    bandwidth = 2.5*10**6
    radar_noise_figure = 6.31
    
    return (p_t * g_t * g_r * lambda_r**2 * sigma) / ((4*math.pi)**3 * radar_range**4 * k_to * bandwidth * radar_noise_figure * loss)

In [5]:
# Call the function below:
# --------------------------

print(signal_to_noise(sigma=3.98, p_t=10**6, g_t=6309.6, g_r=6309.6, radar_range=60*10**3, loss=5.01))

27.406121859256224


**Output:** ~27.41

# Checking conditions:

_In the present section we see how we can use **Conditional statements**, **Logical operators**, **Comparison operators**, **Chained-conditional statements** and **Nested-conditional statements**._

### When to use "if" and "elif"?

_Observe the below code carefully to understanding the difference. Our aim is to check whether a number is divisible by 2 **and** by 3._

In [30]:
# Approach 1:
# ------------

number = 6

if not number%2:
    print("Number divisible by 2")
elif not number%3:
    print("Number divisible by 3")

Number divisible by 2


In [32]:
# Approach 2:
# ------------

number = 6

if not number%2:
    print("Number divisible by 2")
if not number%3:
    print("Number divisible by 3")

Number divisible by 2
Number divisible by 3


_Which of the two approaches do you think is right? and why?_ ;-)

### Task 1:

Define a function called **loan_procurement()** with a parameter called **existing_loan** which takes a boolean as an argument. The function should return **"Loan approved!"** if the customer does not have any existing loan and should return **"Loan rejected"** if the customer has an existing loan.

In [9]:
# Your code below:
# -------------------
# NOTE: (5 min to solve this task)

def loan_procurement(existing_loan):
    if existing_loan:
        return "Loan rejected"
    else:
        return "Loan approved!"

In [10]:
# Check the below conditions:
# -------------------------------

print("You have an existing loan: ", loan_procurement(existing_loan=True))
print("You do not have an existing loan: ", loan_procurement(existing_loan=False))

You have an existing loan:  Loan rejected
You do not have an existing loan:  Loan approved!


**Expected output:**</br>
You have an existing loan:  Loan rejected</br>
You do not have an existing loan:  Loan approved!

### Task 2: 
We will add complexity to the above problem. In the above problem we only checked one condition where a person having an existing loan was rejected a new loan. We introduce one more condition here where we check the person's age. Implement a function based on the below conditions:</br>

**Loan rejected**: If a person has an existing loan </br>

**Loan approved:** If a person has no existing loan and age is below 60 </br>

**Loan under scrutiny:** If a person has no existing loan and age is equal to or above 60 </br>

Please modify the above function to meet these requirements.

In [11]:
# Your code below:
# -------------------
# NOTE: (10 min to solve this task)

def loan_procurement(existing_loan, age):
    if existing_loan:
        return "Loan rejected"
    else:
        if age>=60:
            return "Loan under scrutiny"
        else:
            return "Loan approved!"


In [12]:
# Check the below conditions:
# -------------------------------

print("You have an existing loan and you are 65 years of age: ", loan_procurement(existing_loan=True, age=65))
print("You do not have an existing loan and you are 60 years of age: ", loan_procurement(existing_loan=False, age=60))
print("You do not have an existing loan and you are 67 years of age: ", loan_procurement(existing_loan=False, age=54))

You have an existing loan and you are 65 years of age:  Loan rejected
You do not have an existing loan and you are 60 years of age:  Loan under scrutiny
You do not have an existing loan and you are 67 years of age:  Loan approved!


**Expected output:**</br>
You have an existing loan and you are 65 years of age:  Loan rejected</br>
You do not have an existing loan and you are 60 years of age:  Loan under scrutiny</br>
You do not have an existing loan and you are 67 years of age:  Loan approved!</br>

### Task 3:

We add one additional complexity to the above problem where we add one more condition. We also check the minimum age of a person. If the person is equal to or below 20 years of age, we reject the loan. Modify the function accordingly.

In [13]:
# Your code below:
# -------------------
# NOTE: (5-10 min to solve this task)

def loan_procurement(existing_loan, age):
    if existing_loan or age<=20:
        return "Loan rejected"
    else:
        if age>=60:
            return "Loan under scrutiny"
        else:
            return "Loan approved!"


In [14]:
# Check the below conditions:
# -------------------------------

print("You have an existing loan and you are 65 years of age: ", loan_procurement(existing_loan=True, age=65))
print("You do not have an existing loan and you are 60 years of age: ", loan_procurement(existing_loan=False, age=60))
print("You do not have an existing loan and you are 67 years of age: ", loan_procurement(existing_loan=False, age=54))
print("You do not have an existing loan and you are 16 years of age: ", loan_procurement(existing_loan=False, age=16))

You have an existing loan and you are 65 years of age:  Loan rejected
You do not have an existing loan and you are 60 years of age:  Loan under scrutiny
You do not have an existing loan and you are 67 years of age:  Loan approved!
You do not have an existing loan and you are 16 years of age:  Loan rejected


**Expected output:**</br>
You have an existing loan and you are 65 years of age:  Loan rejected</br>
You do not have an existing loan and you are 60 years of age:  Loan under scrutiny</br>
You do not have an existing loan and you are 67 years of age:  Loan approved!</br>
You do not have an existing loan and you are 16 years of age:  Loan rejected

### Task 4: 

We add one more condition to the above function where we check a person's salary. A loan is rejected if the person earns less than 50,000 Euro. Loan is sent to scrutiny only if your salary is above 80,000 Euro. 

**Note:** Make sure that other conditions from above are satisfied.

In [15]:
# Your code below:
# -------------------
# NOTE: (15 min to solve this task)

def loan_procurement(existing_loan, age, salary):
    if existing_loan or age<=20 or salary<50_000:
        return "Loan rejected"
    else:
        if age<60:
            return "Loan approved!"
        else:
            if salary>80_000:
                return "Loan under scrutiny"
            else:
                return "Loan rejected"


In [16]:
# Check the following conditions:
# ----------------------------------

print("You have an existing loan and you are 25 years old with a salary of 90,000 Euro: ", loan_procurement(existing_loan=True, age=25, salary=90_000))

print("You do not have an existing loan and you are 35 years old with a salary of 90,000 Euro: ", loan_procurement(existing_loan=False, age=35, salary=90_000))

print("You do not have an existing loan and you are 19 years old with a salary of 120,000 Euro: ", loan_procurement(existing_loan=False, age=19, salary=120_000))

print("You do not have an existing loan and you are 65 years old with a salary of 57,000 Euro: ", loan_procurement(existing_loan=False, age=65, salary=57_000))

print("You do not have an existing loan and you are 60 years old with a salary of 81,000 Euro: ", loan_procurement(existing_loan=False, age=60, salary=81_000))

print("You do not have an existing loan and you are 38 years old with a salary of 40,000 Euro: ", loan_procurement(existing_loan=False, age=38, salary=40_000))

print("You do not have an existing loan and you are 35 years old with a salary of 50,000 Euro: ", loan_procurement(existing_loan=False, age=35, salary=50_000))

You have an existing loan and you are 25 years old with a salary of 90,000 Euro:  Loan rejected
You do not have an existing loan and you are 35 years old with a salary of 90,000 Euro:  Loan approved!
You do not have an existing loan and you are 19 years old with a salary of 120,000 Euro:  Loan rejected
You do not have an existing loan and you are 65 years old with a salary of 57,000 Euro:  Loan rejected
You do not have an existing loan and you are 60 years old with a salary of 81,000 Euro:  Loan under scrutiny
You do not have an existing loan and you are 38 years old with a salary of 40,000 Euro:  Loan rejected
You do not have an existing loan and you are 35 years old with a salary of 50,000 Euro:  Loan approved!


**Expected output:** </br>
You have an existing loan and you are 25 years old with a salary of 90,000 Euro:  Loan rejected</br>
You do not have an existing loan and you are 35 years old with a salary of 90,000 Euro:  Loan approved!</br>
You do not have an existing loan and you are 19 years old with a salary of 120,000 Euro:  Loan rejected</br>
You do not have an existing loan and you are 65 years old with a salary of 57,000 Euro:  Loan rejected</br>
You do not have an existing loan and you are 60 years old with a salary of 81,000 Euro:  Loan under scrutiny</br>
You do not have an existing loan and you are 38 years old with a salary of 40,000 Euro:  Loan rejected</br>
You do not have an existing loan and you are 35 years old with a salary of 50,000 Euro:  Loan approved!

### Task 5: 

You allow people to enter your restaurant only if they meet the below conditions. </br>
**Accepted:** Without pets, with reservation</br>
**Rejected:**  With pets, without reservation</br>

On few days, some special considerations are allowed. On Tuesdays you only allow couples to enter. On Wednesdays even pets are allowed. And on Fridays, people can enter without a reservation. Your restaurant is closed on Saturday and Sunday and your function should say it is closed if people come on these days.

Define a function that filters people based on these conditions.


In [17]:
# Your code below:
# -------------------
# NOTE: (15-20 min to solve this task)

def restaurant_filter(day, couple, pets, reservation):
    if day=="Saturday" or day=="Sunday":
        return "Restaurant closed"
    else:
        if day=="Monday" or day=="Thursday":
            if not pets and reservation:
                return "Accepted"
            else: 
                return "Rejected"
        else:
            if day=="Tuesday" and not pets and reservation:
                if couple:
                    return "Accepted"
                else:
                    return "Rejected"
            elif day=="Wednesday" and reservation:
                return "Accepted"
            elif day=="Friday" and not pets:
                return "Accepted"
            else:
                return "Rejected" 
            

In [18]:
# Check the below conditions:
# -------------------------------

# Testing Saturday and Sunday:
print("Testing Saturday and Sunday: ")
print("I am going on a Sunday, with my wife, with a reservation: ", restaurant_filter(day="Sunday", couple=True, pets=False, reservation=True), "\n")

# Testing Monday:
print("Testing Monday: ")
print("I am going on a Monday, with my boy-friend, with a reservation: ", restaurant_filter(day="Monday", couple=True, pets=False, reservation=True))
print("I am going alone on a Monday, with my dog, with a reservation: ", restaurant_filter(day="Monday", couple=False, pets=True, reservation=True), "\n")

# Testing Tuesday:
print("Testing Tuesday: ")
print("I am going on a Tuesday, with my girl-friend, with a reservation: ", restaurant_filter(day="Tuesday", couple=True, pets=False, reservation=True))
print("I am going alone on a Tuesday, with a reservation: ", restaurant_filter(day="Tuesday", couple=False, pets=False, reservation=True), "\n")

# Testing Wednesday:
print("Testing Wednesday: ")
print("I am going alone on a Wednesday, with my cat, with a reservation: ", restaurant_filter(day="Wednesday", couple=False, pets=True, reservation=True))
print("I am going on a Wednesday, with my wife, with a reservation: ", restaurant_filter(day="Wednesday", couple=True, pets=False, reservation=True), "\n")

# Testing Thursday:
print("Testing Thursday: ")
print("I am going on a Thursday, with my donkey, with a reservation: ", restaurant_filter(day="Thursday", couple=False, pets=True, reservation=True))
print("I am going on a Thursday, with my husband, without a reservation: ", restaurant_filter(day="Thursday", couple=True, pets=False, reservation=False), "\n")

# Testing Friday:
print("Testing Friday: ")
print("I am going alone on a Friday, with a reservation: ", restaurant_filter(day="Friday", couple=False, pets=False, reservation=True))
print("I am going on a Friday, with my wife, without a reservation: ", restaurant_filter(day="Friday", couple=True, pets=False, reservation=False), "\n")

Testing Saturday and Sunday: 
I am going on a Sunday, with my wife, with a reservation:  Restaurant closed 

Testing Monday: 
I am going on a Monday, with my boy-friend, with a reservation:  Accepted
I am going alone on a Monday, with my dog, with a reservation:  Rejected 

Testing Tuesday: 
I am going on a Tuesday, with my girl-friend, with a reservation:  Accepted
I am going alone on a Tuesday, with a reservation:  Rejected 

Testing Wednesday: 
I am going alone on a Wednesday, with my cat, with a reservation:  Accepted
I am going on a Wednesday, with my wife, with a reservation:  Accepted 

Testing Thursday: 
I am going on a Thursday, with my donkey, with a reservation:  Rejected
I am going on a Thursday, with my husband, without a reservation:  Rejected 

Testing Friday: 
I am going alone on a Friday, with a reservation:  Accepted
I am going on a Friday, with my wife, without a reservation:  Accepted 



**Expected output:**</br>

Testing Saturday and Sunday: </br>
I am going on a Sunday, with my wife, with a reservation:  Restaurant closed </br>

Testing Monday: </br>
I am going on a Monday, with my boy friend, with a reservation:  Accepted</br>
I am going alone on a Monday, with my dog, with a reservation:  Rejected </br>

Testing Tuesday: </br>
I am going on a Tuesday, with my girl friend, with a reservation:  Accepted</br>
I am going alone on a Tuesday, with a reservation:  Rejected </br>

Testing Wednesday: </br>
I am going alone on a Wednesday, with my cat, with a reservation:  Accepted</br>
I am going on a Wednesday, with my wife, with a reservation:  Accepted </br>

Testing Thursday: </br>
I am going on a Thursday, with my donkey, with a reservation:  Rejected</br>
I am going on a Thursday, with my husband, without a reservation:  Rejected </br>

Testing Friday: </br>
I am going alone on a Friday, with a reservation:  Accepted</br>
I am going on a Friday, with my wife, without a reservation:  Accepted 